package rproto; 


////////////////////////////////////////////////////////
//Types of messages used in the game
enum RMessageType
{
  msg_null = 0;
  msg_shutdown_server = 1;
  msg_chat = 2;
  msg_player_info = 3;
  msg_game_notice = 4;
  msg_srequest = 5;
  msg_crequest = 6;

  msg_lobby_start_game = 7;
  msg_notify_players_loading = 8;
  msg_notify_players_ready = 9;

  msg_map_data = 10;
  msg_ent_data = 11;

  msg_fov_data = 12;

  msg_player_move = 13;
  msg_player_pos = 14;
}


////////////////////////////////////////////////////////
// Chat message
//if it's a normal type then we send it to everyone
//if it's a whisper type then 'reciever_name' will be checked
message ChatMessage
{
  enum ChatType
  {
    chat_null = 0;
    chat_normal = 1;
    chat_whisper = 2;
  }

  optional string owner_ip = 1;
  optional string owner_name = 2;
  optional ChatType msg_type = 3;
  optional string content = 4;

  optional string reciever_name = 5;
}

////////////////////////////////////////////////////////
// Game message, used by server to notify players of things
message GameNotice
{
  enum NoticeType
  {
    m_null = 0;
    m_normal = 1;
    m_error = 2;
    m_notice = 3;
  }

  optional NoticeType msg_type = 1;
  optional string content = 2;
}

////////////////////////////////////////////////////////
//Basic message, used when you just want to send a simple notification
message RMsg
{
  optional RMessageType message_id = 1;
}


////////////////////////////////////////////////////////
// The server uses this to request data from the client
message SRequest
{
  enum RequestType
  {
    r_null = 0;
    r_pinfo = 1;
  }

  optional RequestType request_id = 1;
} 


////////////////////////////////////////////////////////
// The client uses this to request data from the server
message CRequest
{
  enum RequestType
  {
    r_null = 0;
    r_pinfo = 1;
  }

  optional RequestType request_id = 1;
} 


////////////////////////////////////////////////////////
//
message PlayerInfo
{
  optional string name = 1;
  optional string ip = 2;
}


////////////////////////////////////////////////////////
//
message TestMessage
{
  optional string data = 1; 


}




////////////////////////////////////////////////////////
//
message Ent
{
  optional int32 type_id = 1;
  optional int32 game_id = 2;
}

message Cell
{
  optional int32 cx = 1;
  optional int32 cy = 2;
  repeated Ent ents = 3;
}

message View
{
  optional int32 radius = 1;
  repeated Cell cells = 2;
}

////////////////////////////////////////////////////////
//
message PlayerMoveCommand
{
  optional float dx = 1;
  optional float dy = 2;
  optional string name = 3;
}

message PlayerPos
{
  optional float px = 1;
  optional float py = 2;
  optional int32 cx = 3;
  optional int32 cy = 4;
}





























